### PR TITLE
memory_view.c: Add _memory_view_entry member in rb_memory_view_t

### PR DESCRIPTION
**Pull request by mrkn**
To stop looking up the memory_view_entry from the class of the original object of a memory_view in rb_memory_view_release, a pointer of rb_memory_view_entry_t is added as a member of rb_memory_view_t.

This will fix #5012

### CODE REVIEW

**Comment by mrkn** [comment]
`test_memory_view_from_pointer` in `fiddle/test_memory_view.rb` still fails with this change.

```
$ make -C build-debug test-all TESTS='fiddle/test_memory_view.rb -n/test_memory_view_from_pointer/'
make: Entering directory '/home/mrkn/src/github.com/ruby/ruby/build-debug'
Run options:
  --seed=4919
  "--ruby=./miniruby -I../lib -I. -I.ext/common  ../tool/runruby.rb --extout=.ext  -- --disable-gems"
  --excludes-dir=../test/excludes
  --name=!/memory_leak/
  -n/test_memory_view_from_pointer/

# Running tests:

Finished tests in 0.036713s, 27.2387 tests/s, 326.8638 assertions/s.
1 tests, 12 assertions, 0 failures, 0 errors, 0 skips

ruby -v: ruby 3.1.0dev (2021-11-06T02:10:27Z fix_memory_view_re.. b23d6dcf26) [x86_64-linux]
../test/runner.rb: [BUG] Segmentation fault at 0x0000000000000008
ruby 3.1.0dev (2021-11-06T02:10:27Z fix_memory_view_re.. b23d6dcf26) [x86_64-linux]

-- Control frame information -----------------------------------------------
c:0001 p:0000 s:0003 E:0014d0 (none) [FINISH]


-- Machine register context ------------------------------------------------
 RIP: 0x00007f1025f40c06 RBP: 0x00007ffccb8b2c50 RSP: 0x00007ffccb8b2c30
 RAX: 0x0000000000000000 RBX: 0x0000000000000000 RCX: 0x0000000000000000
 RDX: 0x0000000000000000 RDI: 0x00007f1022747910 RSI: 0x0000000000000000
  R8: 0x0000000000000001  R9: 0x0000000000000000 R10: 0xfffffffffffffece
 R11: 0x00007f1025e2b166 R12: 0x0000561d1b4fc160 R13: 0x00007ffccb8b3100
 R14: 0x0000000000000000 R15: 0x0000000000000000 EFL: 0x0000000000010206

-- C level backtrace information -------------------------------------------
/home/mrkn/src/github.com/ruby/ruby/build-debug/libruby.so.3.1.0(rb_print_backtrace+0x1d) [0x7f10260000b5] ../vm_dump.c:759
/home/mrkn/src/github.com/ruby/ruby/build-debug/libruby.so.3.1.0(rb_vm_bugreport+0x15e) [0x7f10260005f3] ../vm_dump.c:1045
/home/mrkn/src/github.com/ruby/ruby/build-debug/libruby.so.3.1.0(rb_bug_for_fatal_signal+0x134) [0x7f1025da6ac7] ../error.c:820
/home/mrkn/src/github.com/ruby/ruby/build-debug/libruby.so.3.1.0(sigsegv+0x71) [0x7f1025f35cde] ../signal.c:964
/home/mrkn/src/github.com/ruby/ruby/build-debug/libruby.so.3.1.0(sigill) (null):0
/lib/x86_64-linux-gnu/libpthread.so.0(__restore_rt+0x0) [0x7f1025c803c0] ../sysdeps/pthread/funlockfile.c:28
/home/mrkn/src/github.com/ruby/ruby/build-debug/libruby.so.3.1.0(do_hash+0x14) [0x7f1025f40c06] ../st.c:320
/home/mrkn/src/github.com/ruby/ruby/build-debug/libruby.so.3.1.0(rb_st_update+0x56) [0x7f1025f429e0] ../st.c:1397
/home/mrkn/src/github.com/ruby/ruby/build-debug/libruby.so.3.1.0(unregister_exported_object+0x56) [0x7f1025e29877] ../memory_view.c:111
/home/mrkn/src/github.com/ruby/ruby/build-debug/libruby.so.3.1.0(rb_memory_view_release+0x6a) [0x7f1025e2b1d0] ../memory_view.c:845
/home/mrkn/src/github.com/ruby/ruby/build-debug/.ext/x86_64-linux/fiddle.so(fiddle_memview_release+0x33) [0x7f10227547e2] ../../../ext/fiddle/memory_view.c:41
/home/mrkn/src/github.com/ruby/ruby/build-debug/.ext/x86_64-linux/fiddle.so(fiddle_memview_free+0x24) [0x7f1022754857] ../../../ext/fiddle/memory_view.c:55
/home/mrkn/src/github.com/ruby/ruby/build-debug/libruby.so.3.1.0(run_final+0x45) [0x7f1025dca090] ../gc.c:4059
/home/mrkn/src/github.com/ruby/ruby/build-debug/libruby.so.3.1.0(finalize_list+0x69) [0x7f1025dca151] ../gc.c:4078
/home/mrkn/src/github.com/ruby/ruby/build-debug/libruby.so.3.1.0(rb_objspace_call_finalizer+0x423) [0x7f1025dca8e0] ../gc.c:4246
/home/mrkn/src/github.com/ruby/ruby/build-debug/libruby.so.3.1.0(rb_ec_finalize+0x3c) [0x7f1025db072f] ../eval.c:164
/home/mrkn/src/github.com/ruby/ruby/build-debug/libruby.so.3.1.0(rb_ec_cleanup+0x44c) [0x7f1025db0bde] ../eval.c:256
/home/mrkn/src/github.com/ruby/ruby/build-debug/libruby.so.3.1.0(ruby_run_node+0x89) [0x7f1025db0f00] ../eval.c:321
/home/mrkn/src/github.com/ruby/ruby/build-debug/ruby(main+0x84) [0x561d1b4fc2cd] ../main.c:47

```

**Comment by kou** [comment]
The following may fix this:

```diff
diff --git a/memory_view.c b/memory_view.c
index b08f13e286..517cd505cb 100644
--- a/memory_view.c
+++ b/memory_view.c
@@ -107,6 +107,8 @@ register_exported_object(VALUE obj)
 static void
 unregister_exported_object(VALUE obj)
 {
+    if (!exported_object_table)
+	return;
     RB_VM_LOCK_ENTER();
     st_update(exported_object_table, (st_data_t)obj, exported_object_dec_ref, 0);
     RB_VM_LOCK_LEAVE();
```

File: memory_view.c [hunk 1]
```diff
@@ -108,6 +108,8 @@ static void
 unregister_exported_object(VALUE obj)
 {
     RB_VM_LOCK_ENTER();
+    if (!exported_object_table)
+        return;
```
Comment (kou) [inline memory_view.c:112]: If we want to do this after `RB_VM_LOCK_ENTER()`, we should call `RUBY_VM_LOCK_LEAVE()` before `return`.

File: memory_view.c [hunk 2]
```diff
@@ -108,6 +108,8 @@ static void
...
     st_update(exported_object_table, (st_data_t)obj, exported_object_dec_ref, 0);
```
Comment (kou) [inline memory_view.c:113]: Or:

```suggestion
    if (exported_object_table)
        st_update(exported_object_table, (st_data_t)obj, exported_object_dec_ref, 0);
```

File: memory_view.c (hunk 1 above)
Comment (mrkn) [reply memory_view.c:112]: OMG! Exactly!

File: memory_view.c (hunk 2 above)
Comment (mrkn) [reply memory_view.c:113]: I like this.

**Pull request merged**